File: Problem_20_Factorial_Digit_Sum.rs
/*
Project: Project Euler - Problem 20: Factorial Digit Sum
Author: Mandeep Bhutani
Date: 10/25/2016

Problem: n! means n * (n-1) * ... * 3 * 2 * 1
For example, 10! = 10 * 9 * ... * 3 * 2 * 1 = 3628800
and the sum of the digits in the number 10! is 3 + 6 + 2 + 8 + 8 + 0 + 0 = 27.
Find the sum of the digits in the number 100!
*/
extern crate num;

use num::Integer;
use num::BigUint;
use num::traits::{One, ToPrimitive};


fn factorial(n: usize) -> BigUint {
    (1..n+1).fold(One::one(), |sum, x| sum * BigUint::from(x))
}


fn sum_digits(n: BigUint) -> u32 {
    let mut num: BigUint = n;
    let mut summation = 0;
    for _ in 1..250 {
        let digit = Integer::mod_floor(&num, &BigUint::from(10 as u32));
        num = Integer::div_floor(&num, &BigUint::from(10 as u32));
        summation += ToPrimitive::to_u32(&digit).unwrap();
    }
    summation
}


fn main() {
    println!("{}", sum_digits(factorial(100)));
}

File: problem_20_factorial_digit_sum.py
"""
Project: Project Euler - Problem 20: Factorial Digit Sum
Author: Mandeep Bhutani
Date: 2/1/2015

Problem:
n! means n * (n-1) * ... * 3 * 2 * 1
For example, 10! = 10 * 9 * ... * 3 * 2 * 1 = 3628800
and the sum of the digits in the number 10! is 3 + 6 + 2 + 8 + 8 + 0 + 0 = 27.
Find the sum of the digits in the number 100!

Analysis: First a function for a factorial needs to be defined.
While there is a math.factorial function that could serve this purpose,
a user defined function was chosen for this problem.
"""


def factorial(n):
    product = 1
    for i in range(2, n+1):
        product *= i
    return product


def sum_factorial_digits(n):
    return sum(int(x) for x in str(factorial(100)))


print(sum_factorial_digits(100))
